core/config/config.hpp

```diff
@@ -0,0 +1,237 @@
+    if (config.is(pnode::status_t::data)) {
+        ptr = context.search_data<LinOpFactory>(config.get_data<std::string>());
+    } else if (config.is(pnode::status_t::map)) {
+        ptr = build_from_config(config, context, exec, td);
+    }
+    // handle object is config
+    assert(ptr.get() != nullptr);
+    return std::move(ptr);
+}
+
+template <>
+std::shared_ptr<const stop::CriterionFactory>
+get_pointer<const stop::CriterionFactory>(const pnode& config,
+                                          const registry& context,
+                                          std::shared_ptr<const Executor> exec,
+                                          type_descriptor td);
+
+
+template <typename T>
+inline std::vector<std::shared_ptr<T>> get_pointer_vector(
+    const pnode& config, const registry& context,
+    std::shared_ptr<const Executor> exec, type_descriptor td)
+{
+    std::vector<std::shared_ptr<T>> res;
+    // for loop in config
+    if (config.is(pnode::status_t::array)) {
+        for (const auto& it : config.get_array()) {
+            res.push_back(get_pointer<T>(it, context, exec, td));
+        }
+    } else {
+        // only one config can be passed without array
+        res.push_back(get_pointer<T>(config, context, exec, td));
+    }
+
+    return res;
+}
+
+template <>
+std::vector<std::shared_ptr<const stop::CriterionFactory>>
+get_pointer_vector<const stop::CriterionFactory>(
+    const pnode& config, const registry& context,
+    std::shared_ptr<const Executor> exec, type_descriptor td);
+
+
+template <typename IndexType>
+inline
+    typename std::enable_if<std::is_integral<IndexType>::value, IndexType>::type
+    get_value(const pnode& config)
+{
+    auto val = config.get_data<long long int>();
+    assert(val <= std::numeric_limits<IndexType>::max() &&
+           val >= std::numeric_limits<IndexType>::min());
+    return static_cast<IndexType>(val);
+}
+
+template <typename ValueType>
+inline typename std::enable_if<std::is_floating_point<ValueType>::value,
+                               ValueType>::type
+get_value(const pnode& config)
+{
+    auto val = config.get_data<double>();
+    assert(val <= std::numeric_limits<ValueType>::max() &&
+           val >= -std::numeric_limits<ValueType>::max());
+    return static_cast<ValueType>(val);
+}
+
+template <typename ValueType>
+inline typename std::enable_if<gko::is_complex_s<ValueType>::value,
+                               ValueType>::type
+get_value(const pnode& config)
+{
+    using real_type = gko::remove_complex<ValueType>;
+    if (config.is(pnode::status_t::data)) {
+        return static_cast<ValueType>(get_value<real_type>(config));
+    } else if (config.is(pnode::status_t::array)) {
+        return ValueType{get_value<real_type>(config.at(0)),
+                         get_value<real_type>(config.at(1))};
+    }
+    GKO_INVALID_STATE("Can not get complex value");
+}
+
+
+#define SET_POINTER(_factory, _param_type, _param_name, _config, _context,     \
```

Comment:
I don't think we use these macros. They save only ~2 lines each. An example of how to replace calls to it can be found here: https://github.com/ginkgo-project/ginkgo/blob/0f459b65535f12d0b43535a13d99ba53aca30adb/core/solver/cg.cpp#L72-L79
With the updates to `property_tree` it would even be a bit shorter. But my point is that we are obfuscating very simple code with this macro.

Reply:
I see your point. For me, macro helps to ensure there's no typo between at("<name>")  and with_<name>, and it's easy to change. for example, changing the `at` and `contain` -> `get` is easy in macro but will be messy in explicit calls.

Reply:
These changes should be fairly easy with non-ancient IDEs. I don't think these are good arguments for using such short macros. Since the code necessary is quite simple, the macros just introduce more confusion IMO.